### DISCUSSION TITLE
Possible to generate 404.html?

### DISCUSSION THREAD

**Discussion by mwskwong** [preactjs/preact-cli#1602] [Q&A]
As suggested by the title, is it possible to generate/pre-render a `404.html` in the production build?

**Comment by rschristian** [comment]
Yep, absolutely. This can be done using a `prerender-urls.json` file. [Docs for which can be found here](https://github.com/preactjs/preact-cli#prerender-multiple-routes)

Take the following, for example:

```json
[
  {
    "url": "/",
    "title": "Homepage"
  },
  {
    "url": "/404",
    "title": "Homepage"
  }
]
```

This will prerender both `/` and `/404`. `/` -> `<build directory>/index.html` and `/404` -> `<build directory>/404/index.html`. The `title` field gets passed into your HTML template, but you can set any arbitrary data this way if you'd like.

**Reply by mwskwong** [reply]
> Yep, absolutely. This can be done using a `prerender-urls.json` file. [Docs for which can be found here](https://github.com/preactjs/preact-cli#prerender-multiple-routes)
> 
> Take the following, for example:
> 
> ```json
> [
>   {
>     "url": "/",
>     "title": "Homepage"
>   },
>   {
>     "url": "/404",
>     "title": "Homepage"
>   }
> ]
> ```
> 
> This will prerender both `/` and `/404`. `/` -> `<build directory>/index.html` and `/404` -> `<build directory>/404/index.html`. The `title` field gets passed into your HTML template, but you can set any arbitrary data this way if you'd like.


The issue is, most hosting platforms will treat `404.html` as the default fallback for 404. And it has to be `404.html` and nothing else. That means I have to add a `postbuild` script to move `/404/index.html` to `/404.html`.

**Reply by rschristian** [reply]
Any static file server should handle that just fine, it's an extremely common pattern and has been for many years.

But yes, if your hosting platform somehow doesn't support serving `/404/index.html` implicitly from `/404` (which sounds seriously odd to me), you'll need to handle that yourself. A `postbuild` script or maybe using `copy-webpack-plugin` in your `preact.config.js`.

**Reply by chickencoding123** [reply]
> Any static file server should handle that just fine, it's an extremely common pattern and has been for many years.

@rschristian that's not true. AWS S3, CloudFront etc... do not understand `404/index.html`. They rely on a root level `404.html` file and `error.html` file when available.

**Reply by rschristian** [reply]
> @rschristian that's not true. AWS S3, CloudFront etc... do not understand `404/index.html`.

Neither of those are web servers so yes, they might not come with all the expected tools out of the box. However, I do see there are configs to allow Cloudfront to support this behavior just fine.
